waku/v2/protocol/waku_peer_exchange/protocol.nim

```diff
@@ -0,0 +1,209 @@
+import
+  std/[options, sets, tables, sequtils, random],
+  stew/results,
+  chronicles,
+  chronos,
+  metrics,
+  libp2p/protocols/protocol,
+  libp2p/crypto/crypto,
+  eth/p2p/discoveryv5/enr
+import
+  ../../node/peer_manager/peer_manager,
+  ../../node/discv5/waku_discv5,
+  ../../utils/requests,
+  ../waku_message,
+  ../waku_relay,
+  ./rpc,
+  ./rpc_codec
+
+
+declarePublicGauge waku_px_peers, "number of peers (in the node's peerManager) supporting the peer exchange protocol"
+declarePublicGauge waku_px_peers_received_total, "number of ENRs received via peer exchange"
+declarePublicGauge waku_px_peers_received_unknown, "number of previously unknown ENRs received via peer exchange"
+declarePublicGauge waku_px_peers_sent, "number of ENRs sent to peer exchange requesters"
+declarePublicGauge waku_px_peers_cached, "number of peer exchange peer ENRs cached"
+declarePublicGauge waku_px_errors, "number of peer exchange errors", ["type"]
+
+logScope:
+  topics = "wakupx"
+
+
+const
+  # We add a 64kB safety buffer for protocol overhead.
+  # 10x-multiplier also for safety
+  MaxRpcSize* = 10 * MaxWakuMessageSize + 64 * 1024 # TODO what is the expected size of a PX message? As currently specified, it can contain an arbitary number of ENRs...
+  MaxCacheSize = 1000
+  CacheCleanWindow = 200
+
+  WakuPeerExchangeCodec* = "/vac/waku/peer-exchange/2.0.0-alpha1"
+
+# Error types (metric label values)
+const
+  dialFailure = "dial_failure"
+  peerNotFoundFailure = "peer_not_found_failure"
+  decodeRpcFailure = "decode_rpc_failure"
+  retrievePeersDiscv5Error= "retrieve_peers_discv5_failure"
+  pxFailure = "px_failure"
+
+type
+  WakuPeerExchangeResult*[T] = Result[T, string]
+
+  WakuPeerExchange* = ref object of LPProtocol
+    peerManager*: PeerManager
+    wakuDiscv5: Option[WakuDiscoveryV5]
+    enrCache: seq[enr.Record] # todo: next step: ring buffer; future: implement cache satisfying https://rfc.vac.dev/spec/34/
+
+proc sendPeerExchangeRpcToPeer(wpx: WakuPeerExchange, rpc: PeerExchangeRpc, peer: RemotePeerInfo | PeerId): Future[WakuPeerExchangeResult[void]] {.async, gcsafe.} =
+  let connOpt = await wpx.peerManager.dialPeer(peer, WakuPeerExchangeCodec)
```

Comment:
I think that without #1161 changes, the `PeerId` version will fail.

Reply:
I think that #1161 is ready to merge, so if we merge the PR before this one, we should be good.

Reply:
Ah, apologies. Fixing signing on #1161 and will merge afterwards. Not sure thought that I see why those changes affect this?

Reply:
#1161 is merged